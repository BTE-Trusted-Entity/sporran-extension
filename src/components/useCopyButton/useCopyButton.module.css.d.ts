declare namespace UseCopyButtonModuleCssNamespace {
  export interface IUseCopyButtonModuleCss {
    copied: string;
    copy: string;
  }
}

declare const UseCopyButtonModuleCssModule: UseCopyButtonModuleCssNamespace.IUseCopyButtonModuleCss & {
  /** WARNING: Only available when `css-loader` is used without `style-loader` or `mini-css-extract-plugin` */
  locals: UseCopyButtonModuleCssNamespace.IUseCopyButtonModuleCss;
};

export = UseCopyButtonModuleCssModule;
