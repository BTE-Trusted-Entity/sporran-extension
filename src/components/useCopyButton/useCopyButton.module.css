.copy {
  composes: buttonIcon from '../Button/Button.module.css';
  background-image: url('../images/copy.svg');
  background-position: center 40%;
  background-size: auto 50%;
  width: 2.25rem;
  height: 2.25rem;
  position: absolute;
  right: 0.25rem;
}

.copied {
  composes: copy;
  background-image: url('../images/checkmark-circle.svg');
}
