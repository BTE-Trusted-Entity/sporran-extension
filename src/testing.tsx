export * from '@testing-library/react';
export { waitForElementToBeRemoved } from '@testing-library/dom';
import { act, render as externalRender, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import dialogPolyfill from 'dialog-polyfill';

import { AccountsProviderMock } from './testing/AccountsProviderMock';

export {
  AccountsProviderMock,
  accountsMock,
} from './testing/AccountsProviderMock';

export { mockBackgroundScript } from './testing/mockBackgroundScript';

export function render(
  ui: Parameters<typeof externalRender>[0],
  options?: Parameters<typeof externalRender>[1],
): ReturnType<typeof externalRender> {
  return externalRender(
    <AccountsProviderMock>
      <MemoryRouter>{ui}</MemoryRouter>
    </AccountsProviderMock>,
    options,
  );
}

export async function waitForTooltipUpdate(): Promise<void> {
  expect(await screen.findByRole('status')).toHaveAttribute(
    'data-popper-escaped',
  );
}

const dialogPromise = Promise.resolve();
(dialogPolyfill.registerDialog as jest.Mock).mockReturnValue(dialogPromise);

export async function waitForDialogUpdate(): Promise<void> {
  await act(() => dialogPromise);
}
