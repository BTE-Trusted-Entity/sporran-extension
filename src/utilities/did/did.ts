import {
  ConfigService,
  Did,
  DidDocument,
  DidEncryptionKey,
  DidUri,
  Utils,
} from '@kiltprotocol/sdk-js';

import { useAsyncValue } from '../useAsyncValue/useAsyncValue';

export function isFullDid(did: DidUri | undefined): boolean {
  if (!did) {
    // could be a legacy identity without DID
    return false;
  }
  return Did.parse(did).type === 'full';
}

export async function getFullDidDocument(did: DidUri): Promise<DidDocument> {
  const api = ConfigService.get('api');
  return Did.linkedInfoFromChain(await api.call.did.query(Did.toChain(did)))
    .document;
}

export function useFullDidDocument(did: DidUri): DidDocument | undefined {
  return useAsyncValue(getFullDidDocument, [did]);
}

export async function getDidDocument(did: DidUri): Promise<DidDocument> {
  return isFullDid(did)
    ? await getFullDidDocument(did)
    : Did.parseDocumentFromLightDid(did);
}

export function getDidEncryptionKey({
  keyAgreement,
}: DidDocument): DidEncryptionKey {
  if (!keyAgreement || !keyAgreement[0]) {
    throw new Error('encryptionKey is not defined somehow');
  }
  return keyAgreement[0];
}

export function parseDidUri(did: DidUri): ReturnType<typeof Did.parse> & {
  fullDid: DidUri;
} {
  const parsed = Did.parse(did);
  const fullDid = parsed.type === 'full' ? did : Did.getFullDidUri(did);

  return {
    ...parsed,
    fullDid,
  };
}

export async function needLegacyDidCrypto(
  did: DidUri | undefined,
): Promise<boolean> {
  if (!did) {
    // DID was deactivated, no action needed.
    return false;
  }
  if (!isFullDid(did)) {
    return false;
  }

  try {
    const encryptionKey = getDidEncryptionKey(await getDidDocument(did));
    return (
      Utils.Crypto.u8aToHex(encryptionKey.publicKey) ===
      '0xf2c90875e0630bd1700412341e5e9339a57d2fefdbba08de1cac8db5b4145f6e'
    );
  } catch {
    // getDidDocument might throw if the DID is not on-chain anymore (removed, another endpoint),
    // no legacy crypto needed in that case
    return false;
  }
}
