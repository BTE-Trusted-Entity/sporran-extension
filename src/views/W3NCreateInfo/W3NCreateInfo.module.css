.container {
  composes: container from '../../components/View/View.module.css';
  height: calc(100% - 3rem);
}

.heading {
  composes: screenHeadline from '../../components/Typography/Typography.module.css';
}

.subline {
  composes: subline from '../../components/Typography/Typography.module.css';
  margin: 0 0 0.8125rem;
}

.didLine {
  composes: didLine from '../../components/CopyValue/CopyValue.module.css';
  margin-top: 1.5rem;
}

.info {
  composes: subline from '../../components/Typography/Typography.module.css';
  line-height: 1.25rem;
}

.warning {
  composes: smallImportant from '../../components/Typography/Typography.module.css';
  width: 100%;
  text-align: left;
  margin-top: 0;
  font-size: 0.875rem;
}

.buttonsLine {
  composes: buttonsLine from '../../components/Button/Button.module.css';
  margin-top: auto;
}

.next {
  composes: buttonPrimary from '../../components/Button/Button.module.css';
}

.back {
  composes: buttonSecondary from '../../components/Button/Button.module.css';
}
