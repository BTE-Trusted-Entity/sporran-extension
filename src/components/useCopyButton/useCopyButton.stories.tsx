import { Meta } from '@storybook/react';
import { useRef } from 'react';

import { useCopyButton } from './useCopyButton';

export default {
  title: 'Components/useCopyButton',
} as Meta;

export function Template(): JSX.Element {
  const ref = useRef(null);
  const copy = useCopyButton(ref);

  return (
    <form>
      <label>
        Input: <input ref={ref} />
      </label>

      {copy.supported && (
        <button
          className={copy.className}
          type="button"
          onClick={copy.handleCopyClick}
          title={copy.title}
          aria-label={copy.title}
        />
      )}
    </form>
  );
}
