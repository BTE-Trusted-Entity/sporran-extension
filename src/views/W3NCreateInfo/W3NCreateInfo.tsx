import { Link, useHistory } from 'react-router-dom';
import { browser } from 'webextension-polyfill-ts';

import * as styles from './W3NCreateInfo.module.css';

import { Identity } from '../../utilities/identities/types';
import { getIdentityDid } from '../../utilities/identities/identities';
import { Stats } from '../../components/Stats/Stats';
import { LinkBack } from '../../components/LinkBack/LinkBack';
import { IdentitySlide } from '../../components/IdentitySlide/IdentitySlide';
import { CopyValue } from '../../components/CopyValue/CopyValue';
import { isFullDid } from '../../utilities/did/did';
import { generatePath, paths } from '../paths';

interface Props {
  identity: Identity;
}

export function W3NCreateInfo({ identity }: Props): JSX.Element {
  const t = browser.i18n.getMessage;
  const { goBack } = useHistory();

  const { address } = identity;
  const did = getIdentityDid(identity);

  const canContinue = isFullDid(did);

  return (
    <section className={styles.container}>
      <h1 className={styles.heading}>{t('view_W3NCreateInfo_heading')}</h1>
      <p className={styles.subline}>{t('view_W3NCreateInfo_subline')}</p>

      <IdentitySlide identity={identity} />

      {canContinue && (
        <CopyValue value={did} label="DID" className={styles.didLine} />
      )}

      <p className={styles.info}>{t('view_W3NCreateInfo_info')}</p>

      {!canContinue && (
        <p className={styles.warning}>{t('view_W3NCreateInfo_warning')}</p>
      )}

      <p className={styles.buttonsLine}>
        <button type="button" onClick={goBack} className={styles.back}>
          {t('common_action_back')}
        </button>

        {canContinue && (
          <Link
            to={generatePath(paths.identity.web3name.create.form, {
              address,
            })}
            className={styles.next}
          >
            {t('common_action_continue')}
          </Link>
        )}
        {!canContinue && (
          <Link
            to={generatePath(paths.identity.did.upgrade.start, { address })}
            className={styles.next}
          >
            {t('view_W3NCreateInfo_upgradeCTA')}
          </Link>
        )}
      </p>

      <LinkBack />
      <Stats />
    </section>
  );
}
