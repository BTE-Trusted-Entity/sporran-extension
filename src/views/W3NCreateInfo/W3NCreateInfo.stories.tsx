import { Meta } from '@storybook/react';

import { identitiesMock } from '../../utilities/identities/IdentitiesProvider.mock';

import { W3NCreateInfo } from './W3NCreateInfo';

export default {
  title: 'Views/W3NCreateInfo',
  component: W3NCreateInfo,
} as Meta;

const on = identitiesMock['4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo'];
const off = identitiesMock['4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1'];

export function OnChainDid(): JSX.Element {
  return <W3NCreateInfo identity={on} />;
}

export function OffChainDid(): JSX.Element {
  return <W3NCreateInfo identity={off} />;
}
