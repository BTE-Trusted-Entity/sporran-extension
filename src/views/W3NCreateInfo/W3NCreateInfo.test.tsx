import { identitiesMock, render } from '../../testing/testing';
import { mockIsFullDid } from '../../utilities/did/did.mock';
import '../../components/useCopyButton/useCopyButton.mock';

import { W3NCreateInfo } from './W3NCreateInfo';

const on = identitiesMock['4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo'];
const off = identitiesMock['4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1'];

describe('W3NCreateInfo', () => {
  it('should match the snapshot for off-chain did', async () => {
    mockIsFullDid(false);
    const { container } = render(<W3NCreateInfo identity={off} />);
    expect(container).toMatchSnapshot();
  });
  it('should match the snapshot for on-chain did', async () => {
    mockIsFullDid(true);
    const { container } = render(<W3NCreateInfo identity={on} />);
    expect(container).toMatchSnapshot();
  });
});
