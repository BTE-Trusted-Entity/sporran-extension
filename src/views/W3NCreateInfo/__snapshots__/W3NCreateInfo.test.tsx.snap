// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`W3NCreateInfo should match the snapshot for off-chain did 1`] = `
<div>
  <section
    class="container"
  >
    <h1
      class="heading"
    >
      Create web3name
    </h1>
    <p
      class="subline"
    >
      Claim a custom name to represent your DID
    </p>
    <section
      class="container"
    >
      <div
        class="avatar"
      >
        Identicon
      </div>
      <div
        class="centeredNameLine"
      >
        <span
          class="name"
        >
          Light DID Identity
        </span>
      </div>
    </section>
    <p
      class="info"
    >
      web3name is a custom name you create to represent your DID and personalize your digital identity. web3names can be linked via your DID with your SocialKYC credentials and with addresses across multiple blockchains.
    </p>
    <p
      class="warning"
    >
      You need a DID to use this feature.
    </p>
    <p
      class="buttonsLine"
    >
      <button
        class="back"
        type="button"
      >
        Back
      </button>
      <a
        class="next"
        href="/identity/4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1/did/upgrade"
      >
        Get your DID
      </a>
    </p>
    <button
      aria-label="Back"
      class="linkBack"
      title="Back"
      type="button"
    />
    <p
      class="stats"
    >
      You have 3 Identities with a
      <span
        class="balance"
      >
         Total Balance: 
        <span
          class="amount"
        >
          3.7020
           
          <span
            aria-label="KILT Coin"
            class="coin"
          />
        </span>
      </span>
    </p>
  </section>
</div>
`;

exports[`W3NCreateInfo should match the snapshot for on-chain did 1`] = `
<div>
  <section
    class="container"
  >
    <h1
      class="heading"
    >
      Create web3name
    </h1>
    <p
      class="subline"
    >
      Claim a custom name to represent your DID
    </p>
    <section
      class="container"
    >
      <div
        class="avatar fullDid"
      >
        Identicon
      </div>
      <div
        class="centeredNameLine"
      >
        <span
          class="name"
        >
          Identity with on-chain DID and lots of coins
        </span>
      </div>
    </section>
    <p
      class="didLine"
    >
      <input
        aria-label="DID"
        class="long"
        readonly=""
        value="did:kilt:4oeJ76hdj84xnwCNqijUHUCTmfwXgSZ4vmxLEiTEYgQdBCcZ"
      />
      <button
        aria-label="copy to clipboard"
        class="copy"
        title="copy to clipboard"
        type="button"
      />
    </p>
    <p
      class="info"
    >
      web3name is a custom name you create to represent your DID and personalize your digital identity. web3names can be linked via your DID with your SocialKYC credentials and with addresses across multiple blockchains.
    </p>
    <p
      class="buttonsLine"
    >
      <button
        class="back"
        type="button"
      >
        Back
      </button>
      <a
        class="next"
        href="/identity/4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo/web3name/create/form"
      >
        Continue
      </a>
    </p>
    <button
      aria-label="Back"
      class="linkBack"
      title="Back"
      type="button"
    />
    <p
      class="stats"
    >
      You have 3 Identities with a
      <span
        class="balance"
      >
         Total Balance: 
        <span
          class="amount"
        >
          3.7020
           
          <span
            aria-label="KILT Coin"
            class="coin"
          />
        </span>
      </span>
    </p>
  </section>
</div>
`;
